crates/bevy_ecs/src/world/mod.rs

```diff
@@ -1653,23 +1662,25 @@ impl World {
     #[inline]
     #[track_caller]
     pub fn init_resource<R: Resource + FromWorld>(&mut self) -> ComponentId {
+        let resource_id = self.register_resource::<R>();
         let caller = MaybeLocation::caller();
-        let component_id = self.components_registrator().register_resource::<R>();
-        if self
-            .storages
-            .resources
-            .get(component_id)
-            .is_none_or(|data| !data.is_present())
-        {
-            let value = R::from_world(self);
-            OwningPtr::make(value, |ptr| {
-                // SAFETY: component_id was just initialized and corresponds to resource of type R.
-                unsafe {
-                    self.insert_resource_by_id(component_id, ptr, caller);
-                }
-            });
+        let resource = R::from_world(self);
```

Comment:
The old code wouldn't call `R::from_world` if the resource already exists, while the new code is calling it unconditionally and relying on `InsertMode::Keep`.  I think we want to avoid running the initializer unnecessarily by checking `entity_mut.contains_id(resource_id)` first.